### PR TITLE
Optimize builds & responsive image generation

### LINKED ISSUE #144: Optimize build speed
Our current implementation generates a Small, Medium and Large image size for each image in our blog. We should drop one of these sizes to improve the total blog size (1.7gb current release). 

- [x] investigate if we should drop small or medium size (looks like dropping small makes the most sense
- [x] implement change 
- [x] validate if latest hugo version is working
- [x] save large pptx as pdf and update blogposts with new link. Solved with #573 and #565
- [ ] fix archetypes; currently each archetype item has an 8mb gif that is unused, should be removed. 
- [ ] Check large images (query for large gifs that could be PNG?)
- [ ] Check current used template metrics

image

### PR DESCRIPTION
**Pull request by andrewconnell**
## Category

- [x] Theme fix

## Related issues

- fixes #144

## Contents of the Pull Request

This PR doubles as an optimization of build time & addressing multiple issues related to how responsive images are generated. Details below in the TL;DR.

Build optimization improvements addressed in two areas:

1. **Heavily leverage Hugo's `partialCached` over `partial`** for elements that are global & not unique to a page such as the sidebar (including categories, tags, and recent posts), footer, CSS, JS, etc. Before, Hugo had to regen these elements for every single page. But if they don't change, just gen them once, cache it, and let Hugo reuse the same element on every page.
1. **Rationalize responsive image generation.** The theme was generating images that were never used on the site. In general, only two sizes are used: mobile & desktop. Rationalized it down from 4 (_including 800w, 1200w & 1500w that were never being used_) to *usually* 2.

    > NOTE on "*usually*": While this slashed ~47% of the image processing, many content pages weren't resizing content images. For example, images referenced as `images/file.png` would get responsively resized, but if referenced as `./images/file.png` wouldn't. Fixing this gap resulted in adding tasks back to image processing. Regardless... appears net reduction still ~29%.

For local build/testing of the site, the first build can take a while (up to 5-10 minutes) to gen all the images involved in image processing (these are created in `./resources/_gen/images`), but subsequent builds are MUCH faster (sub 30s).

👉 Important note about build times... the way the current site updates are configured, it seems like a build can take a long time. Note that BUILDING the site is completely different from DEPLOYING the site. This PR only addresses the BUILD. I believe further optimizations can be made on both the BUILD & DEPLOYMENT aspects, but that's not part of the site content or Hugo theme (what this PR addresses).

Subsequent PR's will address process optimizations around builds & deployments. I need to see this working reliably first before moving onto those.

### TL;DR

Address poor build performance and image responsiveness, including:

- theme image processing issues addressed:
  - was processing images 4x yet 3 of which were never used because the resolution of the generated images is always greater than the max width of the content area... *IOW, these 3 images were garbage tasks*
  - was processing images 4x but without consideration of original image width - for instance, if the source image was 900px, it still resized it UP to 1200px & 1500px which is 100% unnecessary. Only resize down for responsive experiences
  - now images are ONLY resized **down** if source image width > desired image site for scenario
  - instead of 4x images generated for content pages, now max 3x
    - 2x images resized with optimal compression (*.webp)
    - 1x fallback image resized to handle caveman browsers (*.jpeg)
- switch theme partials that are global to all pages to build only once rather than for every page (ie: switch from `partial` to `partialCache`
- fix issue where content pages referencing images as `./images/file.ext` wouldn't get picked up by the image resizer & this potentially result in very large web-unfriendly images on page (image rendering hook now respects `./images/file.ext` & `images/file.ext`

### CODE REVIEW

**Comment by andrewconnell** [comment]
I believe this is ready to go... but with one big caveat: **deployment is a HUGE part time**. The site heavily leverages image processing to create various formats of content images used throughout the site in the interest of responsiveness. **SEE THE PR DESCRIPTION ABOVE FOR MORE DETAIL**.

I have some ideas on cutting the build times down even more, but I need to see this get rolled out.

Cutting down the DEPLOYMENT time is a totally different animal. I'd like to ensure we have a good baseline before proceeding.

**Review by LuiseFreese — APPROVED** [review]
Love this very much. Thanks for your thorough work and documentation. This will bring as a very very big step forward.

**Review by appieschot — APPROVED** [review]

**Comment by appieschot** [comment]
Awesome work 🚀

**Comment by andrewconnell** [comment]
@LuiseFreese said:
> This will bring as a very very big step forward.

Well... let's see it work first :)

**Comment by andrewconnell** [comment]
Do you guys want me to merge this, or do you want to do it? 

Feel free to squash & merge so all the commits don't dirty up the commit history.

**Comment by LuiseFreese** [comment]
#notAGuy but will hit the magic button 😍

**Comment by andrewconnell** [comment]
My bad :) I use "guys" as gender neutral... sort of like "you people" :)
